### ISSUE TITLE
non_blind_divid2k_lmdb pretrain model doesn't use early exit?

### ISSUE THREAD

**Issue by tomcan2015** [RyanXingQL/PowerQE#8]
There are several RBQE pretrained checkpoints provided. For V3, there are many iter_500000.pth files in different subfolders of non_blind_div2k_lmdb\qf20, qf30,qf40,qg50,qp27,qp32,qp37,qp42,qp47. What does the "non_blind" mean? Are these checkpoints use all the layers and don't have the early exit controlled by IQAM?
There are other pretrained checkpoints in V2 and V1 release. Are they use early exit? Which one do you recommend to use?
There is another rbqe project which I guess is the old project and we should use powerqe instead, right?

**Comment by ryanxingql** [comment]
Hi @tomcan2015,

for simplicity, we train only the non-blind version of RBQE in PowerQEV3. That is, all images are exited from the latest exit without early termination. ([README](https://github.com/ryanxingql/powerqe/blob/v3/configs/rbqe/README.md))

This way, the RBQE network outputs only one image, the same as other networks. Therefore, we can use the same training script for all networks in PowerQEV3.

If you want to obtain the early results, then modify the option `if_only_last_output` into `False` ([code](https://github.com/ryanxingql/powerqe/blob/825c41eb910e3630ff85c952e8315553f0bba971/configs/rbqe/non_blind_div2k.py#L16C9-L16C28)). This way, the RBQE network will output all images from all output exits. ([code](https://github.com/ryanxingql/powerqe/blob/825c41eb910e3630ff85c952e8315553f0bba971/powerqe/models/backbones/rbqe.py#L763)) For the training script of this type, please check PowerQEV2 ([config](https://github.com/ryanxingql/powerqe/blob/v2/opts/rbqe.yml)), which details the loss weights of all exits.

**Comment by tomcan2015** [comment]
Thanks for your quick response. For PowerQEV2, there is only one ckp_first_best.pt for jpeg is provided which I guess will use IQAM to determine if we can exit early. 
For PowerQEV3, there are multiple checkpoints are provided. For unknown compression quality input, which checkpoint should we use? In real world, we may have mixed images with different compression QPs. I guess you call this real case as "blind" case which means we don't know the input quality (QP). PowerQEV2 will be the only option to use in this case. Right?
If we know the input quality(QP), then this is a "non-blind" case and we can use PowerQEV3, right?
The Unet structure (channels, kernel size) are the same in V1, V2 and V3?
I did some evaluation of the IQAM block to see if it can properly detect the compression artifacts, but it seems the result doesn't match very well with perceptual evaluation. Image with heavy compression artifacts may get a higher score sometimes. Hope to see some evaluation on this block and compare with other blockiness detection methods. For example, the blockiness detection block used in this paper: "Rethinking Image Super-Resolution from Training Data Perspectives" is working fine.

**Comment by ryanxingql** [comment]
> Thanks for your quick response. For PowerQEV2, there is only one ckp_first_best.pt for jpeg is provided which I guess will use IQAM to determine if we can exit early. For PowerQEV3, there are multiple checkpoints are provided. For unknown compression quality input, which checkpoint should we use? In real world, we may have mixed images with different compression QPs. I guess you call this real case as "blind" case which means we don't know the input quality (QP). PowerQEV2 will be the only option to use in this case. Right? If we know the input quality(QP), then this is a "non-blind" case and we can use PowerQEV3, right? The Unet structure (channels, kernel size) are the same in V1, V2 and V3? I did some evaluation of the IQAM block to see if it can properly detect the compression artifacts, but it seems the result doesn't match very well with perceptual evaluation. Image with heavy compression artifacts may get a higher score sometimes. Hope to see some evaluation on this block and compare with other blockiness detection methods. For example, the blockiness detection block used in this paper: "Rethinking Image Super-Resolution from Training Data Perspectives" is working fine.

If you want to use a pre-trained "blind" model (which mean it is trained with multiple levels), when V2 is the only option. V3 provides different models for different levels.

In fact you can train it on your own data, which should be fast. You need only to modify the training config to assign different weights to different levels of distortions.

We provide two types of IQAM methods, one for HEVC and the other for JPEG. The hyper-parameters of IQAM are tuned on a specific range of QPs/QFs (check paper for details). For your compression, the parameters might need tuning.

**Comment by tomcan2015** [comment]
Thank you again for quick response. One quick question: In your paper: DAQE: Enhancing the Quality of Compressed Images by Exploiting the Inherent Characteristic of Defocus, RBQE result is used to compare with DAQE. For RBQE, do you use powerqe v3 for different QF? "For fair comparisons, we retrain them in a nonblind manner, i.e., train one model for each compression configuration.". It seems you use RBQE trained for QF =20,30,40,50 respectively and all images are exited from the latest exit without early termination.

**Comment by ryanxingql** [comment]
> Thank you again for quick response. One quick question: In your paper: DAQE: Enhancing the Quality of Compressed Images by Exploiting the Inherent Characteristic of Defocus, RBQE result is used to compare with DAQE. For RBQE, do you use powerqe v3 for different QF? "For fair comparisons, we retrain them in a nonblind manner, i.e., train one model for each compression configuration.". It seems you use RBQE trained for QF =20,30,40,50 respectively and all images are exited from the latest exit without early termination.

Hi,
We did use early termination for RBQE. During training, all exits should be trained with the same loss function but different weights; during testing, samples with different QFs are exited in different exits. Then we obtain the enhanced samples.
